src/handle_liquid.cpp

```diff
@@ -460,6 +457,14 @@ bool handle_liquid( item &liquid, const item *const source, const int radius,
         add_msg( _( "The %s froze solid before you could finish." ), liquid.tname() );
         return false;
     }
+    return true;
+}
+
+bool handle_liquid( item &liquid, const item *const source, const int radius,
+                    const tripoint *const source_pos,
+                    const vehicle *const source_veh, const int part_num,
+                    const monster *const source_mon )
+{
```

Comment:
```suggestion
{
    if( !can_handle_liquid( liquid ) ) {
        return false;
    }
```

Reply:
That change would cause the endless loop again this PR aims to fix. The problem is that jumping out of handle_liquid before it has a chance to lower charges results in the loop in 102 to run forever.

Reply:
You already test `can_handle_liquid( liquid )` in the loop condition so there won't be an infinite loop. (If the loop does not terminate, it means `can_handle_liquid( liquid )` must be true, and because `can_handle_liquid` does not modify `liquid`, it is still true here, so this branch is never executed.)